opal/datatype/opal_convertor_internal.h

```diff
@@ -22,10 +23,255 @@
 
 BEGIN_C_DECLS
 
-typedef int32_t (*conversion_fct_t)( opal_convertor_t* pConvertor, uint32_t count,
-                                     const void* from, size_t from_len, ptrdiff_t from_extent,
-                                     void* to, size_t to_length, ptrdiff_t to_extent,
-                                     ptrdiff_t *advance );
+#define COPY_TO_VECTOR   1
+#define COPY_FROM_VECTOR 2
+typedef size_t (*conversion_fct_t)( opal_convertor_t* pConvertor, int mode,
+                                     char* vector_buf,
+                                     size_t vector_len,
+                                     size_t nblocks,
+                                     size_t count_per_block,
+                                     size_t extent_between_blocks,
+                                     size_t vector_element_size,
+                                     int vector_is_bigendian,
+                                     char* packed_buf,
+                                     size_t packed_len,
+                                     size_t packed_element_size,
+                                     int packed_is_bigendian,
+                                     size_t *elements_done );
+
+/*
+ *  The VECTOR_* macros are made to help iterate a buffer pointer over
+ *  data that resembles an MPI_Type_vector, ie it has a number of
+ *  blocks, a count_per_block, an extent_between_blocks, and an element
+ *  size.
+ *
+ *  For example if a buffer's data were
+ *      [INT4,INT4,----,INT4,INT4,----,INT4,INT4]
+ *  the vars describing the vector would be
+ *      nblocks = 3
+ *      count_per_block = 2
+ *      extent_between_blocks = 12
+ *      element_size = 4
+ *
+ *  The macros also have inputs for stopping part of the way through
+ *  a vector, and for picking back up in the middle of the vector.
+ *  The granularity for that is an "element" though, so you could
+ *  stop after writing 3 INT4 in the above example, then pick
+ *  back up at the next INT4, but you can't stop in the middle of
+ *  an INT4.
```

Comment:
True but only for the heterogeneous case. I have tried before to use a 0 trick, to simulate the conversion as a gather/scatter operation, and only move the bytes that are not zero. The code should be still here somewhere using the name _partial_.

Reply:
I didn't really put any work into that.  I just saw the current code didn't seem to have the ability to stop in the middle of a base type so I didn't try to make my code have that ability either.  It doesn't seem like it should be a hard problem if the interface was changed to use bytes for how much iterating had been done, but I'm probably missing something.